InteractionBase/src/vis/Menu.cpp

```diff
@@ -0,0 +1,226 @@
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+#include "Menu.h"
+
+#include "VisualizationBase/src/declarative/DeclarativeItemDef.h"
+#include "VisualizationBase/src/declarative/DynamicGridFormElement.h"
+#include "QGraphicsColorizeEffect"
+#include "VisualizationBase/src/items/Item.h"
+
+namespace Interaction {
+
+ITEM_COMMON_DEFINITIONS(Menu, "item")
+
+Menu* Menu::instance{};
+
+Menu::Menu(QVector<Visualization::Item*> items,
+			Visualization::Item* target, StyleType* style, int nrOfColumns)
+	: Menu(arrange(items, nrOfColumns), target, style)
+{
+}
+
+Menu::Menu(QVector<QVector<Visualization::Item*> > items,
+			Visualization::Item *target, StyleType *style)
+	: Super(nullptr, style), target_(target), currentItems_(items)
+{
+	mousePosition_ = target->scene()->lastMouseHoverPosition();
+	setFiltersChildEvents(true);
+}
+
+Menu::~Menu()
+{
+	//Move the cursor to original owner
+	target_->moveCursor();
+	//Shouldn't be destroyed
+	target_ = nullptr;
+	//Destroyed by DeclarativeItem
+	focusedItem_ = nullptr;
+}
+
+void Menu::hide()
+{
+	if (instance)
+		QApplication::postEvent(instance->scene(),
+								new Visualization::CustomSceneEvent( [&]() { hideNow(); }));
+}
+
+void Menu::hideNow()
+{
+	SAFE_DELETE_ITEM(instance);
+	instance = nullptr;
+}
+
+void Menu::initializeForms()
+{
+	auto nodeGrid = (new Visualization::DynamicGridFormElement())
+			->setSpacing(5, 5)
+			->setMajorAxis(Visualization::GridLayouter::ColumnMajor)
+			->setItemsGetter([](Visualization::Item* v)
+				{ auto self = static_cast<I*>(v);
+				  return self->currentItems(); });
+	//Just have a grid of nodes
+	addForm(nodeGrid);
+}
+
+void Menu::selectItem(Visualization::Item* item)
+{
+	//Make sure the item is part of the menu
+	if (indexOf(item).x() != - 1)
+	{
+		if (focusedItem())
+			focusedItem()->setGraphicsEffect(nullptr);
+		if (item)
+			item->setGraphicsEffect(new QGraphicsColorizeEffect());
+		//Focus the entire menu (somehow just item->moveCursor() doesn't always work)
```

Comment:
It occurs only in an empty ViewItem. In the main ProjectView it seems to work fine without that line.